cc/src/core/internal_contexts.h

```diff
@@ -89,7 +90,10 @@ class PendingContext : public IAsyncContext {
     entry = entry_;
   }
 
-  virtual const key_t& key() const = 0;
+  virtual uint32_t key_size() const = 0;
+  virtual void write_deep_key_at(key_t* dst) const = 0;
```

Comment:
any perf implication of virtual here?

Reply:
The 'virtual' is OK: the sync contexts (e.g. PendingUpsertContext)have the finalized class (with overrides marked as 'final') as template parameter, and the compiler will be able to get rid of the virtual call. The async contexts (e.g. AsyncPendingUpsertContext) will need to go through the virtual call, but the async contexts codepath only triggers when there is a disk IO, so the cost of the virtual call does not matter (the disk IO takes much longer already). 
Also I was not introducing the 'virtual' func thing: the code was already doing this, just look at the line being removed (the virtual const key_t& key() function).